modin/core/storage_formats/base/query_compiler.py

```diff
@@ -363,6 +363,61 @@ def to_numpy(self, **kwargs):  # noqa: PR02
 
     # END To NumPy
 
+    # Dataframe exchange protocol
+
+    def to_dataframe(self, nan_as_null: bool = False, allow_copy: bool = True) -> dict:
```

Comment:
Why not use `@abc.abstractmethod` decorator? How it's done for `to_pandas` method. Same for `from_dataframe`.

Reply:
In order to be able to instantiate a DataFrame with the underlying QC at this line - https://github.com/modin-project/modin/pull/4246/files#diff-028a94e3db5c73c7ddedc6f7b6efb79e07e1360bb1401531b0a9f5d0ca1f1928R55. Otherwise, you will get the error: Can't instantiate abstract class ... with abstract methods from_dataframe, to_dataframe.

Reply:
I see, but isn't that the main way Modin defines an interface?

I propose to define the method as abstract in `BaseQueryCompiler` class, and let the child classes: `PandasQueryCompiler` and `DFAlgQueryCompiler` throw `NotImplementedError` exception.

Reply:
Yes, BaseQueryCompiler should define the interface to be overridden by child classes. We could either make the methods abstract now or as part of efforts discussed [here](https://github.com/modin-project/modin/pull/4226#discussion_r811631015). I don't have strong preferences on this. However, doing that what you proposed

> I propose to define the method as abstract in BaseQueryCompiler class, and let the child classes: PandasQueryCompiler and DFAlgQueryCompiler throw NotImplementedError exception.

You will still be getting the error mentioned above. We should override the methods in TestQC.

Do you think we should make the methods abstract in the Base QC as part of this PR?

Reply:
If there is no hurry, we can do it here. Otherwise, we need to flag this question and implement it elsewhere. (Yes, it also need to de done in `TestQC`)

Reply:
There shouldn't be strong efforts to do that so let's put it here. However, this PR should be merged by end of today.

Reply:
> Yes, it also need to de done in TestQC

That is not quite correct. We should override the methods in TestQC only, PandasQC and OmnisciQC will override the methods as part of new issues.

Reply:
fixed